msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/AcquireTokenSilentlyTest.java

```diff
@@ -116,6 +118,81 @@ void confidentialAppAcquireTokenSilently_claimsSkipCache() throws Throwable {
+
+        //Attempt to retrieve the cached token, however it is expired and should be refreshed.
+        // In this test, it will be replaced with a token that expires in 1 minute
+        responseParameters.put("access_token", "nearlyExpiredToken");
+        responseParameters.put("expires_in", "60");
+        TestHelper.createTokenRequestMock(httpClientMock, TestHelper.getSuccessfulTokenResponse(responseParameters), 200);
+
+        SilentParameters silentParameters = SilentParameters.builder(Collections.singleton("someScopes"), result.account()).build();
+        result = cca.acquireTokenSilently(silentParameters).get();
+
+        //Ensure there is still one token in the cache, however it is the new refreshed token rather than the token from the first mocked call
+        assertRefreshedToken(result, "nearlyExpiredToken", CacheRefreshReason.EXPIRED, cca.tokenCache.accessTokens.size());
+
+        //Attempt to retrieve the cached token, however it is within the 5-minute buffer and should be refreshed.
+        // In this test, it will be replaced with a token that expires in 1 hour
+        responseParameters.put("access_token", "normalToken");
+        responseParameters.put("expires_in", "3600");
+        TestHelper.createTokenRequestMock(httpClientMock, TestHelper.getSuccessfulTokenResponse(responseParameters), 200);
+
+        silentParameters = SilentParameters.builder(Collections.singleton("someScopes"), result.account()).build();
+        result = cca.acquireTokenSilently(silentParameters).get();
+
+        assertRefreshedToken(result, "normalToken", CacheRefreshReason.EXPIRED, cca.tokenCache.accessTokens.size());
+
+        //Force the token to be refreshed
+        responseParameters.put("access_token", "forcedRefreshToken");
+        TestHelper.createTokenRequestMock(httpClientMock, TestHelper.getSuccessfulTokenResponse(responseParameters), 200);
+
+        silentParameters = SilentParameters.builder(Collections.singleton("someScopes"), result.account()).forceRefresh(true).build();
+        result = cca.acquireTokenSilently(silentParameters).get();
+
+        assertRefreshedToken(result, "forcedRefreshToken", CacheRefreshReason.FORCE_REFRESH, cca.tokenCache.accessTokens.size());
+
+        //Finally, force a refresh by setting claims
+        responseParameters.put("access_token", "claimsToken");
+        TestHelper.createTokenRequestMock(httpClientMock, TestHelper.getSuccessfulTokenResponse(responseParameters), 200);
+
+        silentParameters = SilentParameters.builder(Collections.singleton("someScopes"), result.account()).claims(new ClaimsRequest()).build();
+        result = cca.acquireTokenSilently(silentParameters).get();
+
+        assertRefreshedToken(result, "claimsToken", CacheRefreshReason.CLAIMS, cca.tokenCache.accessTokens.size());
+    }
+
+    //Asserts that there is one expected token in the cache, and that it was refreshed with the expected reason
```

Comment:
can you also add a test for proactive token renewal?

Reply:
Added in the latest commit.